manticore/native/models.py

```diff
@@ -140,3 +144,96 @@ def strlen(state, s):
             ret = ITEBV(cpu.address_bit_size, byt == 0, offset, ret)
 
     return ret
+
+
+def is_definitely_NULL(byte, constrs) -> bool:
+    """
+    Checks if a given byte read from memory is NULL.
+    This supports both concrete & symbolic byte values.
+
+    :param byte: byte read from memory to be examined
+    :param constrs: state constraints
+    :return: whether a given byte is NULL or constrained to NULL
+    """
+    if issymbolic(byte):
+        return Z3Solver.instance().must_be_true(constrs, byte == 0)
+    else:
+        return byte == 0
+
+
+def cannot_be_NULL(byte, constrs) -> bool:
+    """
+    Checks if a given byte read from memory is not NULL or cannot be NULL
+
+    :param byte: byte read from memory to be examined
+    :param constrs: state constraints
+    :return: whether a given byte is not NULL or cannot be NULL
+    """
+    if issymbolic(byte):
+        return Z3Solver.instance().must_be_true(constrs, byte != 0)
+    else:
+        return byte != 0
+
+
+def can_be_NULL(byte, constrs) -> bool:
+    """
+    Checks if a given byte read from memory can be NULL
+
+    :param byte: byte read from memory to be examined
+    :param constrs: state constraints
+    :return: whether a given byte is NULL or can be NULL
+    """
+    if issymbolic(byte):
+        return Z3Solver.instance().can_be_true(constrs, byte == 0)
+    else:
+        return byte == 0
+
+
+def strcpy(state: State, dst: Union[int, BitVec], src: Union[int, BitVec]) -> Union[int, BitVec]:
+    """
+    strcpy symbolic model
+
+    Algorithm: Copy every byte from src to dst until finding a byte that is NULL or is
+    constrained to only the NULL value. Every time a byte is fouund that can be NULL but
+    is not definetly NULL concretize and fork states.
+
+    :param state: current program state
+    :param dst: destination string address
+    :param src: source string address
+    :return: pointer to the dst
+    """
+    if issymbolic(src):
+        raise ConcretizeArgument(state.cpu, 1)
+
+    if issymbolic(dst):
+        raise ConcretizeArgument(state.cpu, 0)
+
+    cpu = state.cpu
+    constrs = state.constraints
+    ret = dst
+
+    # Initialize offset based on whether state has been forked in strcpy
+    if "strcpy" not in state.context:
+        offset = 0
+    else:
+        offset = state.context["strcpy"]
+
+    # Copy until a src_byte is symbolic and constrained to '\000', or is concrete and '\000'
+    src_val = cpu.read_int(src + offset, 8)
+    while not is_definitely_NULL(src_val, constrs):
+        cpu.write_int(dst + offset, src_val, 8)
+
+        # If a byte can be NULL set the src_val for concretize and fork states
+        if can_be_NULL(src_val, constrs):
+            state.context["strcpy"] = offset
```

Comment:
As a note for myself, this is okay because state contexts aren't locked like plugin contexts are. It's one context per state, and each state's context isn't accessible from other threads, so we shouldn't get weird issues with states stepping on one another.